# Day 1: Handling missing values

This procedure is based on blogs available on internet:

1) https://www.kaggle.com/rtatman/data-cleaning-challenge-handling-missing-values

2) https://towardsdatascience.com/the-complete-beginners-guide-to-data-cleaning-and-preprocessing-2070b7d4c6d

3) https://elitedatascience.com/data-cleaning

4) https://www.kaggle.com/getting-started/125899#729278

Most of this notebook is based on 4. The dataset "San Francisco Building Permits" is used, available in Kaggle https://www.kaggle.com/aparnashastry/building-permit-applications-data#DataDictionaryBuildingPermit.xlsx.

In this tutorial, some "quick and dirty" techniques were covered. This can help with missing values but will probably also end up removing some useful information or adding some noise to the data.

### 1) Imports first!

In [1]:
import pandas as pd
import numpy as np

In [4]:
#We read the data
sf_permits = pd.read_csv("Dataset/Building_Permits.csv")

/Users/ricardorodriguez/anaconda3/envs/DeepLearning/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3051: DtypeWarning: Columns (22,32) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [23]:
sf_permits.shape

(198900, 43)

In [5]:
#We can observe the first elements of the dataframe to have an idea of the dataset
sf_permits.head()

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0           05/06/2015  0326  023            140                  NaN   
1           04/19/2016  0306  007            440                  NaN   
2           05/27/2016  0595  203           1647                  NaN   
3           11/07/2016  0156  011           1230                  NaN   
4           11/28/2016  0342  001            950                  NaN   

  Street Name Street Suffix  ...  Existing Construction Type  \
0       Ellis            St  ...                         3.0   
1       Geary            St  ...                         3.0   
2     Pacific            Av  ...                         1.0   
3     Pacific            Av  ...                         5.0   
4      Market            St  ...                         3.0   

  Existing Construction Type Description Proposed Construction Type  \
0                          constr type 3                        NaN   
1                          constr type 3                        NaN   
2                          constr type 1                        1.0   
3                         wood frame (5)                        5.0   
4                          constr type 3                        NaN   

  Proposed Construction Type Description Site Permit Supervisor District  \
0                                    NaN         NaN                 3.0   
1                                    NaN         NaN                 3.0   
2                          constr type 1         NaN                 3.0   
3                         wood frame (5)         NaN                 3.0   
4                                    NaN         NaN                 6.0   

  Neighborhoods - Analysis Boundaries  Zipcode  \
0                          Tenderloin  94102.0   
1                          Tenderloin  94102.0   
2                        Russian Hill  94109.0   
3                            Nob Hill  94109.0   
4                          Tenderloin  94102.0   

                                    Location      Record ID  
0  (37.785719256680785, -122.40852313194863)  1380611233945  
1   (37.78733980600732, -122.41063199757738)  1420164406718  
2    (37.7946573324287, -122.42232562979227)  1424856504716  
3   (37.79595867909168, -122.41557405519474)  1443574295566  
4   (37.78315261897309, -122.40950883997789)   144548169992  

[5 rows x 43 columns]

In [8]:
#However, we can take select a random portion of the datafram for a better understanding of the dataset
np.random.seed(0) 
sf_permits.sample(5)

Permit Number  Permit Type            Permit Type Definition  \
40553   201403039652            8            otc alterations permit   
169731  201510159735            3  additions alterations or repairs   
19180        M409787            8            otc alterations permit   
68047   201411191888            8            otc alterations permit   
64238        M527228            8            otc alterations permit   

       Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
40553            03/03/2014  3732  008            400                  NaN   
169731           10/15/2015  2609  028             79                  NaN   
19180            07/22/2013  4624  031            178                  NaN   
68047            11/19/2014  0039  109            294                  NaN   
64238            10/14/2014  1251  002            707                  NaN   

        Street Name Street Suffix  ...  Existing Construction Type  \
40553    Clementina            St  ...                         NaN   
169731  Buena Vista            Tr  ...                         5.0   
19180    West Point            Rd  ...                         NaN   
68047     Francisco            St  ...                         5.0   
64238          Cole            St  ...                         NaN   

       Existing Construction Type Description Proposed Construction Type  \
40553                                     NaN                        1.0   
169731                         wood frame (5)                        5.0   
19180                                     NaN                        NaN   
68047                          wood frame (5)                        5.0   
64238                                     NaN                        NaN   

       Proposed Construction Type Description Site Permit Supervisor District  \
40553                           constr type 1         NaN                 6.0   
169731                         wood frame (5)         NaN                 8.0   
19180                                     NaN         NaN                10.0   
68047                          wood frame (5)         NaN                 3.0   
64238                                     NaN         NaN                 5.0   

       Neighborhoods - Analysis Boundaries  Zipcode  \
40553                      South of Market  94103.0   
169731                 Castro/Upper Market  94117.0   
19180                Bayview Hunters Point  94124.0   
68047                          North Beach  94133.0   
64238                       Haight Ashbury  94117.0   

                                         Location      Record ID  
40553   (37.780460571778164, -122.40450626524974)  1334094491645  
169731   (37.76757916496494, -122.43793170417105)  1399356139170  
19180    (37.73524725436046, -122.38063828309745)  1311685491725  
68047   (37.805257822817126, -122.40998545760392)  1362881288870  
64238    (37.76836885973765, -122.45074431487859)   135886493776  

[5 rows x 43 columns]

### 2) Check for missing values

From the previous observation, we observe that the dataset contains 43 columns. It is pretty clear that we have some missing values in some columns. Now, we will see with more detail the number of missing values. 

In [19]:
# get the number of missing data points per column
missing_values_count = sf_permits.isnull().sum()

# look at the # of missing points in the first twenty columns
missing_values_count

Permit Number                                  0
Permit Type                                    0
Permit Type Definition                         0
Permit Creation Date                           0
Block                                          0
Lot                                            0
Street Number                                  0
Street Number Suffix                      196684
Street Name                                    0
Street Suffix                               2768
Unit                                      169421
Unit Suffix                               196939
Description                                  290
Current Status                                 0
Current Status Date                            0
Filed Date                                     0
Issued Date                                14940
Completed Date                            101709
First Construction Document Date           14946
Structural Notification                   191978
Number of Existing S

Apparently there exists many missing cells. Let's observe the percentage of missing data.

In [16]:
# how many total missing values do we have?
total_cells = np.product(sf_permits.shape)
total_missing = missing_values_count.sum()

# percent of data that is missing
(total_missing/total_cells) * 100

26.26002315058403

We have 26% of cells with missing data, slightly more than a quarter of the cells. In order to deal with this missing data, we must understand why is it missing.

### 3) Figure out why the data is missing - Data intuition

We have to check wheter this value is missing because it wasn't recorded or because it dosen't exist. Should we do some "imputation" for missing values?

Let's take a look into the 20 first columns

In [21]:
missing_values_count[0:20]

Permit Number                            0
Permit Type                              0
Permit Type Definition                   0
Permit Creation Date                     0
Block                                    0
Lot                                      0
Street Number                            0
Street Number Suffix                196684
Street Name                              0
Street Suffix                         2768
Unit                                169421
Unit Suffix                         196939
Description                            290
Current Status                           0
Current Status Date                      0
Filed Date                               0
Issued Date                          14940
Completed Date                      101709
First Construction Document Date     14946
Structural Notification             191978
dtype: int64

We see many columns that can have relevant information. To follow the tutorial, let's think about the columns <b>Street Number Suffix </b> and <b> Zipcode </b>. 

By looking at the documentation, 'Street Number Suffix' is a value "Related to address". This column has a really large number of missing values (196684). There are a very few values that exists for some specific rows, only 1%. So, they might be missing because they don't exist at all. On the otherside, 'Zipcode' (1716) has a few rows where the values are missing so, it's probably because it wasn't recorded.

### 4) Option 1: Drop missing values

Note: This approch is not recommended for important projects! It's usually worth it to take the time to go through your data and really look at all the columns with missing values one-by-one to really get to know your dataset

In [31]:
# remove all the rows that contain a missing value
sf_permits.dropna()

Empty DataFrame
Columns: [Permit Number, Permit Type, Permit Type Definition, Permit Creation Date, Block, Lot, Street Number, Street Number Suffix, Street Name, Street Suffix, Unit, Unit Suffix, Description, Current Status, Current Status Date, Filed Date, Issued Date, Completed Date, First Construction Document Date, Structural Notification, Number of Existing Stories, Number of Proposed Stories, Voluntary Soft-Story Retrofit, Fire Only Permit, Permit Expiration Date, Estimated Cost, Revised Cost, Existing Use, Existing Units, Proposed Use, Proposed Units, Plansets, TIDF Compliance, Existing Construction Type, Existing Construction Type Description, Proposed Construction Type, Proposed Construction Type Description, Site Permit, Supervisor District, Neighborhoods - Analysis Boundaries, Zipcode, Location, Record ID]
Index: []

[0 rows x 43 columns]

All the rows has gone since each row in our dataset contains at least one Nan value. We might try to remove all the columns that have at least one missing value instead.

In [32]:
# remove all columns with at least one missing value
columns_with_na_dropped = sf_permits.dropna(axis=1)
columns_with_na_dropped.head()

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   

  Permit Creation Date Block  Lot  Street Number Street Name Current Status  \
0           05/06/2015  0326  023            140       Ellis        expired   
1           04/19/2016  0306  007            440       Geary         issued   
2           05/27/2016  0595  203           1647     Pacific      withdrawn   
3           11/07/2016  0156  011           1230     Pacific       complete   
4           11/28/2016  0342  001            950      Market         issued   

  Current Status Date  Filed Date      Record ID  
0          12/21/2017  05/06/2015  1380611233945  
1          08/03/2017  04/19/2016  1420164406718  
2          09/26/2017  05/27/2016  1424856504716  
3          07/24/2017  11/07/2016  1443574295566  
4          12/01/2017  11/28/2016   144548169992

In [34]:
# just how much data did we lose?
print("Columns in original dataset: %d \n" % sf_permits.shape[1])
print("Columns with na's dropped: %d" % columns_with_na_dropped.shape[1])

Columns in original dataset: 43 

Columns with na's dropped: 12


We've lost quite a bit of data, but at this point we have successfully removed all the NaN's from our data.

### 5) Option 2 - Filling in missing values automatically

Filling in missing values is also known as "imputation"

In [36]:
# get a small subset of the NFL dataset
subset_permits_data = sf_permits.loc[:, 'Permit Number':'Completed Date'].head()
subset_permits_data

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0           05/06/2015  0326  023            140                  NaN   
1           04/19/2016  0306  007            440                  NaN   
2           05/27/2016  0595  203           1647                  NaN   
3           11/07/2016  0156  011           1230                  NaN   
4           11/28/2016  0342  001            950                  NaN   

  Street Name Street Suffix  Unit Unit Suffix  \
0       Ellis            St   NaN         NaN   
1       Geary            St   0.0         NaN   
2     Pacific            Av   NaN         NaN   
3     Pacific            Av   0.0         NaN   
4      Market            St   NaN         NaN   

                                         Description Current Status  \
0  ground fl facade: to erect illuminated, electr...        expired   
1            remove (e) awning and associated signs.         issued   
2                    installation of separating wall      withdrawn   
3           repair dryrot & stucco at front of bldg.       complete   
4  demolish retail/office/commercial 3-story buil...         issued   

  Current Status Date  Filed Date Issued Date Completed Date  
0          12/21/2017  05/06/2015  11/09/2015            NaN  
1          08/03/2017  04/19/2016  08/03/2017            NaN  
2          09/26/2017  05/27/2016         NaN            NaN  
3          07/24/2017  11/07/2016  07/18/2017     07/24/2017  
4          12/01/2017  11/28/2016  12/01/2017            NaN

One option to fill the data is to set all the missing values by one value. We can try with 0 for instance.

In [37]:
subset_permits_data.fillna(0)

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   

  Permit Creation Date Block  Lot  Street Number  Street Number Suffix  \
0           05/06/2015  0326  023            140                     0   
1           04/19/2016  0306  007            440                     0   
2           05/27/2016  0595  203           1647                     0   
3           11/07/2016  0156  011           1230                     0   
4           11/28/2016  0342  001            950                     0   

  Street Name Street Suffix  Unit  Unit Suffix  \
0       Ellis            St   0.0            0   
1       Geary            St   0.0            0   
2     Pacific            Av   0.0            0   
3     Pacific            Av   0.0            0   
4      Market            St   0.0            0   

                                         Description Current Status  \
0  ground fl facade: to erect illuminated, electr...        expired   
1            remove (e) awning and associated signs.         issued   
2                    installation of separating wall      withdrawn   
3           repair dryrot & stucco at front of bldg.       complete   
4  demolish retail/office/commercial 3-story buil...         issued   

  Current Status Date  Filed Date Issued Date Completed Date  
0          12/21/2017  05/06/2015  11/09/2015              0  
1          08/03/2017  04/19/2016  08/03/2017              0  
2          09/26/2017  05/27/2016           0              0  
3          07/24/2017  11/07/2016  07/18/2017     07/24/2017  
4          12/01/2017  11/28/2016  12/01/2017              0

Another option is to replace missing values with whatever value comes directly after it in the same column. (This makes a lot of sense for datasets where the observations have some sort of logical order to them.)

In [38]:
# replace all NA's the value that comes directly after it in the same column, 
# then replace all the reamining na's with 0
subset_permits_data.fillna(method = 'bfill', axis=0).fillna(0)

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   

  Permit Creation Date Block  Lot  Street Number  Street Number Suffix  \
0           05/06/2015  0326  023            140                   0.0   
1           04/19/2016  0306  007            440                   0.0   
2           05/27/2016  0595  203           1647                   0.0   
3           11/07/2016  0156  011           1230                   0.0   
4           11/28/2016  0342  001            950                   0.0   

  Street Name Street Suffix  Unit  Unit Suffix  \
0       Ellis            St   0.0          0.0   
1       Geary            St   0.0          0.0   
2     Pacific            Av   0.0          0.0   
3     Pacific            Av   0.0          0.0   
4      Market            St   0.0          0.0   

                                         Description Current Status  \
0  ground fl facade: to erect illuminated, electr...        expired   
1            remove (e) awning and associated signs.         issued   
2                    installation of separating wall      withdrawn   
3           repair dryrot & stucco at front of bldg.       complete   
4  demolish retail/office/commercial 3-story buil...         issued   

  Current Status Date  Filed Date Issued Date Completed Date  
0          12/21/2017  05/06/2015  11/09/2015     07/24/2017  
1          08/03/2017  04/19/2016  08/03/2017     07/24/2017  
2          09/26/2017  05/27/2016  07/18/2017     07/24/2017  
3          07/24/2017  11/07/2016  07/18/2017     07/24/2017  
4          12/01/2017  11/28/2016  12/01/2017              0